File: workout_service/src/test/java/com/louisfiges/workout/validation/ValidatorTest.java
package com.louisfiges.workout.validation;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ValidatorTest {

    private final Validator validator = new Validator(3, 64);

    @Test
    void reportsSpecificReasonsForShortStrings() {
        assertThat(validator.isInvalidString("Ab")).isTrue();
        assertThat(validator.getStringValidationMessage("Workout name", "Ab"))
                .isEqualTo("Workout name must be at least 3 characters long");
    }

    @Test
    void reportsWhitespaceAndCharacterProblemsTogether() {
        assertThat(validator.getStringValidationIssues(" #"))
                .containsExactly(
                        "leading_or_trailing_whitespace",
                        "too_short",
                        "invalid_characters"
                );

        assertThat(validator.getStringValidationMessage("Exercise name", " #"))
                .isEqualTo(
                        "Exercise name must not start or end with whitespace, " +
                                "be at least 3 characters long, and " +
                                "contain only letters, numbers, spaces, and common punctuation"
                );
    }

    @Test
    void reportsRequiredStrings() {
        assertThat(validator.getStringValidationMessage("Workout type", ""))
                .isEqualTo("Workout type is required");
        assertThat(validator.isInvalidString("")).isTrue();
    }
}

File: frontend/tests/unit/utils/validator.test.ts
import { describe, expect, it } from "vitest";
import {
  getStringValidationIssues,
  getStringValidationMessage,
  isInvalidString,
} from "@/utils/validator";

describe("validator string helpers", () => {
  it("flags short strings with a specific reason", () => {
    expect(getStringValidationIssues("Ab")).toContain("too_short");
    expect(getStringValidationMessage("Workout name", "Ab")).toContain(
      "at least",
    );
    expect(isInvalidString("Ab")).toBe(true);
  });

  it("reports whitespace and character problems together", () => {
    expect(getStringValidationIssues(" #")).toEqual(
      expect.arrayContaining([
        "leading_or_trailing_whitespace",
        "too_short",
        "invalid_characters",
      ]),
    );
    expect(getStringValidationMessage("Exercise name", " #")).toContain(
      "not start or end with whitespace",
    );
    expect(getStringValidationMessage("Exercise name", " #")).toContain(
      "contain only letters, numbers, spaces, and common punctuation",
    );
  });
});
